# Logistic Regression
This programm runs logistic regression in order to predict wether stock is a good option to buy. Stock is classified as 'Buy' if it'll beat SP500 and ROI is above 2%.

## First attempt (fewer data)

### 1. Imports

In [8]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler

### 2. Load the data

In [11]:
data = pd.read_csv('stocks_data.csv')
data.describe(include='all')

Unnamed: 0 Ticker          Year         Month      MA Ratio  \
count   14854.000000  14854  14854.000000  14854.000000  14854.000000   
unique           NaN    393           NaN           NaN           NaN   
top              NaN    AWK           NaN           NaN           NaN   
freq             NaN     55           NaN           NaN           NaN   
mean     7426.500000    NaN   2020.669113      6.223240      1.004148   
std      4288.124784    NaN      1.428016      3.520757      0.046473   
min         0.000000    NaN   2018.000000      1.000000      0.580721   
25%      3713.250000    NaN   2019.000000      3.000000      0.977766   
50%      7426.500000    NaN   2021.000000      6.000000      1.005360   
75%     11139.750000    NaN   2022.000000      9.000000      1.031953   
max     14853.000000    NaN   2023.000000     12.000000      1.335656   

                 Buy        Result           ROE  Insider Ownership Growth  \
count   14854.000000  14854.000000  14854.000000              14854.000000   
unique           NaN           NaN           NaN                       NaN   
top              NaN           NaN           NaN                       NaN   
freq             NaN           NaN           NaN                       NaN   
mean        0.415848      1.032817     39.494365                  0.015486   
std         0.492884      0.149260    181.839873                  0.269863   
min         0.000000      0.259712   -613.743387                 -0.633527   
25%         0.000000      0.944153     10.160854                 -0.001350   
50%         0.000000      1.028547     19.251991                  0.000000   
75%         1.000000      1.113949     31.949569                  0.008000   
max         1.000000      3.142137  11300.330487                 17.759731   

        Institutional Ownership Growth  Forecast EPS Growth  \
count                     14854.000000         14854.000000   
unique                             NaN                  NaN   
top                                NaN                  NaN   
freq                               NaN                  NaN   
mean                          0.026708             0.057775   
std                           0.230675             2.136724   
min                          -0.714136            -0.992366   
25%                          -0.023114            -0.184264   
50%                          -0.000648            -0.039062   
75%                           0.033653             0.086957   
max                          14.203540           133.000000   

        Avg 2Q EPS Growth  Avg 2Q EPS Surprise  YoY EPS Growth  \
count        14854.000000         14854.000000    14854.000000   
unique                NaN                  NaN             NaN   
top                   NaN                  NaN             NaN   
freq                  NaN                  NaN             NaN   
mean             0.181477            13.755183        0.369529   
std              2.111809            46.751483        3.637998   
min            -45.050000           -65.625000       -0.961538   
25%             -0.040838             2.015000        0.017606   
50%              0.045662             6.055000        0.130688   
75%              0.154182            13.135000        0.275148   
max             78.553797          1493.215000      192.000000   

        Sector Performance  Market Performance  Benchmark SP500 Performance  
count         14854.000000        14854.000000                 14854.000000  
unique                 NaN                 NaN                          NaN  
top                    NaN                 NaN                          NaN  
freq                   NaN                 NaN                          NaN  
mean              1.488003            1.438443                     1.029219  
std               8.164589            7.038394                     0.077491  
min             -44.900728          -22.795349                     0.769903  
25%              -3.45

### 3. Clean the data
Let's remove outliers.

In [12]:
for column in ['ROE', 'Insider Ownership Growth', 'Institutional Ownership Growth', 'Forecast EPS Growth', 'Avg 2Q EPS Growth', 'YoY EPS Growth']:
    upper_bound = data[column].quantile(0.999)
    data = data[(data[column] <= upper_bound)]
q = data['Avg 2Q EPS Growth'].quantile(0.001)
data = data[(data['Avg 2Q EPS Growth'] >= q)]
data.describe(include='all')

Unnamed: 0 Ticker          Year         Month      MA Ratio  \
count   14757.000000  14757  14757.000000  14757.000000  14757.000000   
unique           NaN    391           NaN           NaN           NaN   
top              NaN    IQV           NaN           NaN           NaN   
freq             NaN     55           NaN           NaN           NaN   
mean     7431.564546    NaN   2020.665447      6.222877      1.004135   
std      4285.343004    NaN      1.428264      3.521282      0.046414   
min         0.000000    NaN   2018.000000      1.000000      0.580721   
25%      3726.000000    NaN   2019.000000      3.000000      0.977791   
50%      7430.000000    NaN   2021.000000      6.000000      1.005388   
75%     11144.000000    NaN   2022.000000      9.000000      1.031914   
max     14853.000000    NaN   2023.000000     12.000000      1.335656   

                 Buy        Result           ROE  Insider Ownership Growth  \
count   14757.000000  14757.000000  14757.000000              14757.000000   
unique           NaN           NaN           NaN                       NaN   
top              NaN           NaN           NaN                       NaN   
freq             NaN           NaN           NaN                       NaN   
mean        0.415735      1.032886     35.930619                  0.009596   
std         0.492865      0.149125     99.720733                  0.072866   
min         0.000000      0.259712   -613.743387                 -0.633527   
25%         0.000000      0.944257     10.176162                 -0.001361   
50%         0.000000      1.028536     19.293997                  0.000000   
75%         1.000000      1.113877     31.949569                  0.007968   
max         1.000000      3.142137   1723.426083                  1.636364   

        Institutional Ownership Growth  Forecast EPS Growth  \
count                     14757.000000         14757.000000   
unique                             NaN                  NaN   
top                                NaN                  NaN   
freq                               NaN                  NaN   
mean                          0.021883             0.011317   
std                           0.133643             0.519399   
min                          -0.714136            -0.992366   
25%                          -0.023167            -0.182927   
50%                          -0.000699            -0.038835   
75%                           0.033287             0.086705   
max                           1.596845             8.514706   

        Avg 2Q EPS Growth  Avg 2Q EPS Surprise  YoY EPS Growth  \
count        14757.000000         14757.000000    14757.000000   
unique                NaN                  NaN             NaN   
top                   NaN                  NaN             NaN   
freq                  NaN                  NaN             NaN   
mean             0.145736            13.522146        0.272787   
std              0.806091            45.365794        1.069267   
min             -3.489011           -65.625000       -0.961538   
25%             -0.040646             1.995000        0.018152   
50%              0.045532             6.010000        0.130396   
75%              0.152364            13.080000        0.274611   
max             27.731828          1493.215000       31.500000   

        Sector Performance  Market Performance  Benchmark SP500 Performance  
count         14757.000000        14757.000000                 14757.000000  
unique                 NaN                 NaN                          NaN  
top                    NaN                 NaN                          NaN  
freq                   NaN                 NaN                          NaN  
mean              1.486683            1.441679                     1.029255  
std               8.159292            7.043774                     0.077495  
min             -44.900728          -22.795349                     0.769903  
25%              -3.45

### 4. Split the data for train and test, standarise the data

In [13]:
data = data.reset_index(drop=True)
train_data = data[data['Year'] <= 2022]
test_data = data[data['Year'] > 2022]
x_train = train_data.drop(['Year', 'Buy', 'Month', 'Ticker', 'Result', 'Benchmark SP500 Performance', data.columns[0]], axis=1)
y_train = train_data['Buy']
x_test = test_data.drop(['Year', 'Buy', 'Month', 'Ticker', 'Result', 'Benchmark SP500 Performance', data.columns[0]], axis=1)
y_test = test_data['Buy']

scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)

x_test = scaler.transform(x_test)

### 5. Train the model

In [79]:
model = LogisticRegression(max_iter = 10000, class_weight={0: 100, 1: 100})
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

### 6. Evaluation

In [80]:
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
precision = precision_score(y_test, y_pred)
print(f'Precision: {precision}')

conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.6596664607782582
Precision: 0.28
Confusion Matrix:
[[1061   18]
 [ 533    7]]
Classification Report:
              precision    recall  f1-score   support

           0       0.67      0.98      0.79      1079
           1       0.28      0.01      0.02       540

    accuracy                           0.66      1619
   macro avg       0.47      0.50      0.41      1619
weighted avg       0.54      0.66      0.54      1619



In [81]:
test_results = test_data.copy()
test_results['Predicted_Buy'] = y_pred

predicted_stocks_to_buy = test_results[test_results['Predicted_Buy'] == 1]
predicted_stocks_avg_return = predicted_stocks_to_buy['Result'].mean()

avg_stock_return = test_results['Result'].mean()

best_stocks = test_results[test_results['Buy'] == 1]
avg_best_stocks_return = best_stocks['Result'].mean()

sp500_return = predicted_stocks_to_buy['Benchmark SP500 Performance'].mean()

print("Benchmarks: ")
print(f"Average stock return (whole test sample): {avg_stock_return:.3f}")
print(f"'Buy' stocks average return: {avg_best_stocks_return:.3f}")
print(f"SP500 return: {sp500_return:.3f}")

print(f"\nModel's predicted stock average return: {predicted_stocks_avg_return:.3f}")

Benchmarks: 
Average stock return (whole test sample): 1.032
'Buy' stocks average return: 1.176
SP500 return: 1.057

Model's predicted stock average return: 1.043


### 7. Conclusion
The results achieved by the model leave much to be desired. Both the low precision and poorer performance compared to the S&P 500 suggest that logistic regression may not be suitable for building an investment strategy in this case. To investigate this further, more data is needed as well as the application of feature engineering. For example, by modifying the class weights and giving 20% more weight to class 0, the model achieves an ROI of 13%, which is a very good result compared to the S&P 500 (5%). However, the recall is extremely low, below 1%, so there is a high probability that this is just a fluke.

## Second attempt (more data)
In this attempt, data with a broader date range (from around 2008 for most companies) was used. This range includes several recession periods, providing greater diversity for the model to learn from. Additionally, columns containing information on the companies' Return on Assets and Return on Invested Capital were added.

### 1. Read & preprocess the data

In [107]:
data = pd.read_csv('stocks_data3.csv')
data.describe(include='all')

Unnamed: 0 Ticker          Year         Month      MA Ratio  \
count   61517.000000  61517  61517.000000  61517.000000  61517.000000   
unique           NaN    417           NaN           NaN           NaN   
top              NaN    CPB           NaN           NaN           NaN   
freq             NaN    225           NaN           NaN           NaN   
mean    30758.000000    NaN   2015.561064      6.475576      1.005282   
std     17758.572592    NaN      5.130421      3.448068      0.043844   
min         0.000000    NaN   2005.000000      1.000000      0.580721   
25%     15379.000000    NaN   2012.000000      3.000000      0.982233   
50%     30758.000000    NaN   2016.000000      6.000000      1.006824   
75%     46137.000000    NaN   2020.000000      9.000000      1.029836   
max     61516.000000    NaN   2024.000000     12.000000      1.443446   

                 Buy        Result           ROE           ROA           ROI  \
count   61517.000000  61517.000000  61517.000000  61517.000000  61517.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.454102      1.035127      0.262087      0.089610      0.150720   
std         0.497893      0.139926      8.540118      1.104599      1.285496   
min         0.000000      0.110349   -347.693570     -1.369770    -15.336400   
25%         0.000000      0.955759      0.095910      0.037540      0.065280   
50%         0.000000      1.035871      0.166400      0.070180      0.114670   
75%         1.000000      1.113424      0.266340      0.113230      0.186960   
max         1.000000      3.538378    509.372300    157.982210    181.327650   

        Insider Ownership Growth  Institutional Ownership Growth  \
count               61517.000000                    61517.000000   
unique                       NaN                             NaN   
top                          NaN                             NaN   
freq                         NaN                             NaN   
mean                    0.031192                        0.025757   
std                     0.891108                        0.250149   
min                    -0.994779                       -0.930676   
25%                    -0.003494                       -0.020228   
50%                     0.000000                        0.000790   
75%                     0.008696                        0.027434   
max                    76.000000                       14.203540   

        Forecast EPS Growth  Avg 2Q EPS Growth  Avg 2Q EPS Surprise  \
count          61517.000000       61517.000000         61517.000000   
unique                  NaN                NaN                  NaN   
top                     NaN                NaN                  NaN   
freq                    NaN                NaN                  NaN   
mean               0.090401           0.149241            11.124411   
std                1.732801           1.515826            52.925722   
min               -0.992366         -58.668103           -93.235000   
25%               -0.156716          -0.036162             1.270000   
50%               -0.016129           0.047591             4.740000   
75%                0.110000           0.157784            10.680000   
max              133.000000          78.553797          3750.995000   

        YoY EPS Growth  Sector Performance  Market Performance  \
count     6.151700e+04        61517.000000        61517.000000   
unique             NaN                 NaN                 NaN   
top                NaN                 NaN                 NaN   
freq               NaN                 NaN                 NaN   
mean      3.969275e+12            1.641091            1.603703   
std       4.301144e+14            7.192112            6.071008   
min      -1.000000e+00          -49.5014

In [108]:
for column in ['ROE', 'Insider Ownership Growth', 'Institutional Ownership Growth', 'Forecast EPS Growth', 'Avg 2Q EPS Growth', 'YoY EPS Growth']:
    upper_bound = data[column].quantile(0.999)
    data = data[(data[column] <= upper_bound)]
q = data['Avg 2Q EPS Growth'].quantile(0.001)
data = data[(data['Avg 2Q EPS Growth'] >= q)]
data.describe(include='all')

Unnamed: 0 Ticker          Year         Month      MA Ratio  \
count   61087.000000  61087  61087.000000  61087.000000  61087.000000   
unique           NaN    417           NaN           NaN           NaN   
top              NaN    SHW           NaN           NaN           NaN   
freq             NaN    225           NaN           NaN           NaN   
mean    30760.834187    NaN   2015.569221      6.478662      1.005279   
std     17750.016433    NaN      5.121289      3.449648      0.043759   
min         0.000000    NaN   2005.000000      1.000000      0.580721   
25%     15398.500000    NaN   2012.000000      3.000000      0.982252   
50%     30760.000000    NaN   2016.000000      6.000000      1.006816   
75%     46143.500000    NaN   2020.000000      9.000000      1.029818   
max     61516.000000    NaN   2024.000000     12.000000      1.443446   

                 Buy        Result           ROE           ROA           ROI  \
count   61087.000000  61087.000000  61087.000000  61087.000000  61087.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.453894      1.035114      0.111004      0.082006      0.141937   
std         0.497874      0.139492      4.354249      0.065388      0.227534   
min         0.000000      0.110349   -347.693570     -1.369770    -15.336400   
25%         0.000000      0.955893      0.096130      0.037730      0.065425   
50%         0.000000      1.035912      0.166670      0.070250      0.114810   
75%         1.000000      1.113338      0.266075      0.113290      0.186910   
max         1.000000      3.538378     19.528570      1.313780     23.687470   

        Insider Ownership Growth  Institutional Ownership Growth  \
count               61087.000000                    61087.000000   
unique                       NaN                             NaN   
top                          NaN                             NaN   
freq                         NaN                             NaN   
mean                    0.011203                        0.020213   
std                     0.160279                        0.158886   
min                    -0.994779                       -0.930676   
25%                    -0.003509                       -0.020261   
50%                     0.000000                        0.000723   
75%                     0.008621                        0.027216   
max                     4.750000                        3.115815   

        Forecast EPS Growth  Avg 2Q EPS Growth  Avg 2Q EPS Surprise  \
count          61087.000000       61087.000000         61087.000000   
unique                  NaN                NaN                  NaN   
top                     NaN                NaN                  NaN   
freq                    NaN                NaN                  NaN   
mean               0.048165           0.135944            10.844773   
std                0.661588           0.638213            45.682348   
min               -0.992366          -7.247826           -93.235000   
25%               -0.156250          -0.035667             1.275000   
50%               -0.016129           0.047619             4.730000   
75%                0.108752           0.157197            10.642500   
max               16.888889          16.014706          2768.315000   

        YoY EPS Growth  Sector Performance  Market Performance  \
count     61087.000000        61087.000000        61087.000000   
unique             NaN                 NaN                 NaN   
top                NaN                 NaN                 NaN   
freq               NaN                 NaN                 NaN   
mean          0.204320            1.642579            1.606065   
std           0.708250            7.179243            6.067306   
min          -1.000000          -49.5014

### 2. Split the data and train the model

In [111]:
cut_off_year = 2019

data = data.reset_index(drop=True)
train_data = data[(data['Year'] < cut_off_year) & ((data['Year'] != cut_off_year - 1) | (data['Month'] < 9))]
test_data = data[data['Year'] >= cut_off_year]
x_train = train_data.drop(['Year', 'Buy', 'Month', 'Ticker', 'Result', 'Benchmark SP500 Performance', data.columns[0]], axis=1)
y_train = train_data['Buy']
x_test = test_data.drop(['Year', 'Buy', 'Month', 'Ticker', 'Result', 'Benchmark SP500 Performance', data.columns[0]], axis=1)
y_test = test_data['Buy']

scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)

x_test = scaler.transform(x_test)

In [112]:
model = LogisticRegression(max_iter = 10000, class_weight={0: 150, 1: 100})
model.fit(x_train, y_train)
y_pred = model.predict(x_test)

### 3. Evaluate the mode;

In [113]:
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')
precision = precision_score(y_test, y_pred)
print(f'Precision: {precision}')

conf_matrix = confusion_matrix(y_test, y_pred)
print('Confusion Matrix:')
print(conf_matrix)

class_report = classification_report(y_test, y_pred)
print('Classification Report:')
print(class_report)

Accuracy: 0.5961421319796955
Precision: 0.5425531914893617
Confusion Matrix:
[[11693    43]
 [ 7913    51]]
Classification Report:
              precision    recall  f1-score   support

           0       0.60      1.00      0.75     11736
           1       0.54      0.01      0.01      7964

    accuracy                           0.60     19700
   macro avg       0.57      0.50      0.38     19700
weighted avg       0.57      0.60      0.45     19700



In [114]:
test_results = test_data.copy()
test_results['Predicted_Buy'] = y_pred

predicted_stocks_to_buy = test_results[test_results['Predicted_Buy'] == 1]
predicted_stocks_avg_return = predicted_stocks_to_buy['Result'].mean()

avg_stock_return = test_results['Result'].mean()

best_stocks = test_results[test_results['Buy'] == 1]
avg_best_stocks_return = best_stocks['Result'].mean()

sp500_return = predicted_stocks_to_buy['Benchmark SP500 Performance'].mean()

print("Benchmarks: ")
print(f"Average stock return (whole test sample): {avg_stock_return:.3f}")
print(f"'Buy' stocks average return: {avg_best_stocks_return:.3f}")
print(f"SP500 return: {sp500_return:.3f}")

print(f"\nModel's predicted stock average return: {predicted_stocks_avg_return:.3f}")

Benchmarks: 
Average stock return (whole test sample): 1.034
'Buy' stocks average return: 1.160
SP500 return: 1.057

Model's predicted stock average return: 1.122


### 4. Conclusions
The model performs significantly better after being retrained on a larger dataset. Despite a modest accuracy of 51%, the model achieved an average ROI of 11.8% per 'Buy' signal. This improvement underscores the positive impact of increasing the dataset size on the model's accuracy and predictive capabilities. Such a result suggests that the model can be used to build a profitable investment strategy, but it also requires other elements, such as risk management and a precise definition of entry and exit strategies.